### ISSUE TITLE
Telescope-style search box

### ISSUE THREAD

**Issue by SalamanderXing** [zed-industries/zed#8279]
### Check for existing issues

- [X] Completed

### Describe the feature

In (neo)vim I'm able to search for file names and contents using telescope. This has the following advantages:
-  runs search **as I type** 
- I don't have to leave the current tab so it's faster
- Since the preview of a file and the list of files are separated I can just very quickly scroll through the list of files only using my keyboard

### If applicable, add mockups / screenshots to help present your vision of the feature

<img width="1908" alt="Screenshot 2024-02-23 at 15 57 27" src="https://github.com/zed-industries/zed/assets/6129632/89e29b58-ac2c-4668-8833-9c3e3136bf54">

**Comment by ryanwinchester** [comment]
It's the main thing I'll miss from neovim

<details><summary>video</summary>

https://github.com/zed-industries/zed/assets/2897340/f0bb89f5-43eb-4e1e-9c1b-dd8214aec952

</details>

And the grep version

<details><summary>grep video</summary>

https://github.com/zed-industries/zed/assets/2897340/bfe1e504-13f2-46ba-bea0-4df3d67d6fb5

</details>

**Comment by ryanwinchester** [comment]
And the grep version

https://github.com/zed-industries/zed/assets/2897340/bfe1e504-13f2-46ba-bea0-4df3d67d6fb5

**Comment by NachoNievaG** [comment]
Agree completely.
Also, the live grep requires [ripgrep](https://github.com/BurntSushi/ripgrep) (rust based) as a dependency, which using it in Zed sounds suitable. (talking as an outsider regarding integrations tho)

**Comment by lungarella-raffaele** [comment]
Would be the feature that makes me move from vs-code

**Comment by martincartledge** [comment]
Please make this happen!

**Comment by joachimhorshauge** [comment]
This would be amazing

**Comment by ElhamAryanpur** [comment]
absolutely, I'm on board too

**Comment by zeystiv** [comment]
I also want this to happen, It'll speed the coding.

**Comment by costinsin** [comment]
This would be very nice!

**Comment by vjdhanota** [comment]
In case it helps anyone else, I was able to set up a task that uses `rg`, `fzf` and `bat` for a better find in files experience:

```
  {
    "label": "search:project",
    "command": "QUERY=${ZED_SELECTED_TEXT:-}; FILE=$(rg --column --hidden --line-number --no-heading --color=always --smart-case --colors match:fg:green --colors path:fg:white --colors path:style:nobold --glob '!**/.git/' '--glob' '!**/node_modules'  \"$QUERY\" . | fzf --ansi --delimiter : --preview 'bat --style=numbers,changes,header --color=always --highlight-line {2} {1}' --preview-window 'up:60%:+{2}+3/3' --layout=reverse --query \"$QUERY\" --print-query | tail -1); if [ -n \"$FILE\" ]; then LINE=$(echo $FILE | cut -d':' -f2); COL=$(echo $FILE | cut -d':' -f3); FILE=$(echo $FILE | cut -d':' -f1); zed $FILE:$LINE:$COL; fi",
    "env": {},
    "cwd": "${ZED_WORKTREE_ROOT}",
    "allow_concurrent_runs": false,
    "use_new_terminal": false,
    "reveal": "always",
    "hide": "on_success"
  },
```

The preview window is specific to my layout but should be pretty easy to update (`--preview-window 'up:60%:+{2}+3/3' --layout=reverse`). There is probably a cleaner way to set this up but it has worked for me so far.

Something built-in like telescope would be amazing

**Comment by j-o-sh** [comment]
This 🙌

I'm soooo in favor. This is the main (and atm. basically the only) missing feature that keeps me from switching to Zed from Neovim. - I don't care for tabs or clunky file trees that much but am totally content using Telescope for this.

This I'd need from that:
- Fuzzy find files in the root dir (basic Telescope)
- Grep over files in root dir
- Find files in the open buffers

Bonus:
- Grep over files in the open buffers
- Find (and Grep in) files that have been modified in the current branch of the repo

**Comment by Koshkaj** [comment]
If this gets added to the editor along with errorlens. consider me as a loyal lifelong Zed user.

**Comment by eduDorus** [comment]
This would be a very nice addition! Keep up the good work

**Comment by dkuku** [comment]
Every time I try to switch from nvim I end up just running nvim in the editor terminal to edit files this way :D
For me this is a killer feature.

**Comment by calebeaires** [comment]
This is the final piece to make me migrate from webstorm to zed

**Comment by Rits1272** [comment]
Planning to move from neo vim to zed. Wish this gets implemented in zed

**Comment by bauffman** [comment]
I've been trying Zed for a few weeks now and the only feature that keeps me from making the switch is Telescope. Hope this gets released and keep up the good work!

**Comment by GengCen-Qin** [comment]
> In case it helps anyone else, I was able to set up a task that uses `rg`, `fzf` and `bat` for a better find in files experience:
> 
> ```
>   {
>     "label": "search:project",
>     "command": "QUERY=${ZED_SELECTED_TEXT:-}; FILE=$(rg --column --hidden --line-number --no-heading --color=always --smart-case --colors match:fg:green --colors path:fg:white --colors path:style:nobold --glob '!**/.git/' '--glob' '!**/node_modules'  \"$QUERY\" . | fzf --ansi --delimiter : --preview 'bat --style=numbers,changes,header --color=always --highlight-line {2} {1}' --preview-window 'up:60%:+{2}+3/3' --layout=reverse --query \"$QUERY\" --print-query | tail -1); if [ -n \"$FILE\" ]; then LINE=$(echo $FILE | cut -d':' -f2); COL=$(echo $FILE | cut -d':' -f3); FILE=$(echo $FILE | cut -d':' -f1); zed $FILE:$LINE:$COL; fi",
>     "env": {},
>     "cwd": "${ZED_WORKTREE_ROOT}",
>     "allow_concurrent_runs": false,
>     "use_new_terminal": false,
>     "reveal": "always",
>     "hide": "on_success"
>   },
> ```
> 
> The preview window is specific to my layout but should be pretty easy to update (`--preview-window 'up:60%:+{2}+3/3' --layout=reverse`). There is probably a cleaner way to set this up but it has worked for me so far.
> 
> Something built-in like telescope would be amazing

Thank you for providing the code solution. I have adjusted it to make it look even better. Everyone is welcome to give it a try.

https://github.com/user-attachments/assets/bcef53da-5229-47eb-99b8-b09ead7f1723

> my fish config: 

```sh
function search_and_edit
    set -l selected_file (rg --column --hidden --line-number --no-heading --color=always --smart-case --glob '!**/.git/' --glob '!**/node_modules/' . | fzf --ansi --delimiter : --preview 'bat --style=numbers,changes,header --color=always --highlight-line {2} {1}' --preview-window 'up:60%:+{2}+3/3' --layout=reverse)

    if test -n "$selected_file"
        set -l file (echo $selected_file | cut -d':' -f1)
        set -l line (echo $selected_file | cut -d':' -f2)
        set -l col (echo $selected_file | cut -d':' -f3)

        zed $file:$line:$col
    end
end
```

> my keymap

```json
{
    "context": "Editor && vim_mode == normal && !VimWaiting && !menu",
    "bindings": {
      "space f": [
        "workspace::SendKeystrokes",
        ": new center terminal enter search_and_edit space && space exit enter"
      ]
    }
  }
```

**Comment by JaLnYn** [comment]
Hey guys, I was looking at this issue too. I have an implementation of this search style in rust on my websocket ide project [here](https://github.com/JaLnYn/websocket-ide/blob/main/src/search/search_manager.rs) on line 177. I use the crate nucleo. 

I will spend some time looking into the code to see if I can easily move the coded over. 

If someone can point to a starting point, that would be great.

**Comment by JaLnYn** [comment]
I submitted a pr for realtime search

https://github.com/zed-industries/zed/pull/21348

but it's kinda slow because their current search seems to lack optimizations. Waiting on their advice on how to proceed.

**Comment by LogicKahanHai** [comment]
Hey @GengCen-Qin Can you please help me out.. I configured fish and added the function as well and it works in the terminal when I put it manually. I have also added the keymap as you provided. But, when i press `space-f` it just opens the full-screen terminal tab for me and then stops there. I am unable to find any solution to this. Would appreciate your help. Thanks :)

**Comment by vjdhanota** [comment]
The `"workspace::SendKeystrokes"` workaround keybinding to open a center terminal no longer works as of a recent release. Luckily, the latest preview release added support for running tasks in a center pane (`Zed Preview 0.167.1`). 

I've updated and enhanced my config since my last post. I've added file search and historical search support. Sharing it here for others:

In my 

<details><summary>`~/.zshrc`</summary>

```sh
function zed_file_search() {
  local FILE
  FILE=$(
    rg --files --hidden --color=always \
      --glob '!**/.git/' \
      --glob '!**/node_modules' \
    | fzf --ansi --preview 'bat --decorations=always --color=always {} --style=full --line-range :50' \
          --preview-window 'up:60%:wrap' \
          --layout=reverse
  )

  if [ -n "$FILE" ]; then
    zed "$FILE"
  fi
}

function zed_search() {
  local query="${1:-}"
  local OUT

  # Run fzf in multi-selection mode with dynamic reload based on the query.
  # The first line of output is the final query, followed by one or more selected results.
  OUT=$(
    fzf --ansi --cycle --exact --delimiter ':' --nth 4.. \
      --preview 'bat --decorations=always --color=always {1} --highlight-line {2} --style=full' \
      --preview-window 'up:60%:+{2}+3/3' \
      --layout=reverse \
      --query "$query" \
      --phony \
      --print-query \
      --multi \
      --bind 'start:reload:rg --column --hidden --line-number --no-heading --color=always --smart-case \
        --colors match:fg:green --colors path:fg:white --colors path:style:nobold \
        --glob "!**/.git/" --glob "!**/node_modules" {q} .' \
      --bind 'change:reload:rg --column --hidden --line-number --no-heading --color=always --smart-case \
        --colors match:fg:green --colors path:fg:white --colors path:style:nobold \
        --glob "!**/.git/" --glob "!**/node_modules" {q} .'
  )

  # If nothing was selected, do nothing
  if [ -z "$OUT" ]; then
    return
  fi

  # The first line of OUT is the query used, append it to history
  local FINAL_QUERY
  FINAL_QUERY=$(echo "$OUT" | head -n1)

  if [ -n "$FINAL_QUERY" ]; then
    echo "$FINAL_QUERY" >> ~/.zed_search_history
  fi

  # Now extract the selected file lines (everything after the first line)
  # Each line is in the format: filepath:line:col:...
  echo "$OUT" | tail -n +2 | while IFS= read -r FILE_LINE; do
    if [ -n "$FILE_LINE" ]; then
      local LINE COL PATH
      # Not sure why, but I have to use absolute paths to `cut` and `zed` here
      LINE=$(echo "$FILE_LINE" | /usr/bin/cut -d':' -f2)
      COL=$(echo "$FILE_LINE" | /usr/bin/cut -d':' -f3)
      PATH=$(echo "$FILE_LINE" | /usr/bin/cut -d':' -f1)
      /usr/local/bin/zed "$PATH:$LINE:$COL"
    fi
  done
}

function zed_search_history() {
  # FZF to choose a past query
  local HIST_QUERY=$(fzf --prompt="History> " < ~/.zed_search_history)
  if [ -n "$HIST_QUERY" ]; then
    # Run zed_search again with the chosen query
    zed_search "$HIST_QUERY"
  fi
}
```

</details>

<details><summary>My global `tasks.json`</summary>

```jsonc
  {
    "label": "search:files",
    "command": "zed_file_search",
    "env": {},
    "cwd": "${ZED_WORKTREE_ROOT}",
    "allow_concurrent_runs": false,
    "use_new_terminal": false,
    "reveal": "always",
    "hide": "always",
    "show_command": true,
    "show_summary": true,
    "reveal_target": "center"
  },
  {
    "label": "search:project",
    "command": "zed_search ${ZED_SELECTED_TEXT:-}",
    "env": {},
    "cwd": "${ZED_WORKTREE_ROOT}",
    "allow_concurrent_runs": false,
    "use_new_terminal": false,
    "reveal": "always",
    "hide": "always",
    "show_command": true,
    "show_summary": true,
    "reveal_target": "center"
  },
  {
    "label": "search:history",
    "command": "zed_search_history",
    "env": {},
    "cwd": "${ZED_WORKTREE_ROOT}",
    "allow_concurrent_runs": false,
    "use_new_terminal": false,
    "reveal": "always",
    "hide": "always",
    "show_command": true,
    "show_summary": true,
    "reveal_target": "center"
  },
```

</details>

I then have some keybindings that fire these tasks. Here is a quick demo of how it works/looks:

<details><summary>video</summary>

https://github.com/user-attachments/assets/8438ed15-aa6d-4b89-953a-758581e05915

</details>

I recently came across this TUI tool for general-purpose fuzzy finding inspired by Telescope - https://github.com/alexpasmantier/television. Integrating this as a center terminal task might be a better approach but I haven't gotten around to it yet

**Comment by baldwindavid** [comment]
@vjdhanota - Nice work on that and good find for `television`. Just a naive setup leveraging that utility works well enough for me. I'm sure this could be improved, but doesn't get much simpler.

`tasks.json`

```json
  {
    "label": "File Finder",
    "command": "zed \"$(tv files)\"",
    "hide": "always",
    "allow_concurrent_runs": true,
    "use_new_terminal": true
  },
  {
    "label": "Find in Files",
    "command": "zed \"$(tv text)\"",
    "hide": "always",
    "allow_concurrent_runs": true,
    "use_new_terminal": true
  },
```

`keymap.json`

```json
"cmd-p": [
  "task::Spawn",
  { "task_name": "File Finder", "reveal_target": "center" }
],
"cmd-f": [
  "task::Spawn",
  { "task_name": "Find in Files", "reveal_target": "center" }
],
```

**Comment by LogicKahanHai** [comment]
Hi @baldwindavid 
I tried your method. Amazing! Thank you so much for this! Just one small question, how does the `reveal_target: center` part of it work (or affect the whole thing) I took a look at #9445 and #22004 and from there what I understood was that adding this would run the task in the center terminal (like the one that opens when you run `: new center terminal`) but it does not do that in my system. Did I miss something in between? I really want it to run in the center terminal. 


Thanks for your help! Merry Christmas!

**Comment by baldwindavid** [comment]
@LogicKahanHai The feature to run a task in the center terminal is currently only in the [preview release](https://zed.dev/releases/preview). The next stable release will include it.

**Comment by baldwindavid** [comment]
Posted a discussion of the drop-in replacement leveraging television with a video in #22581. This is effectively a solved issue for me at this point though I know popping into a TUI is not everyone's bag.

**Comment by hassanu-t** [comment]
This would be the feature that really makes the move for me from nvim to zed!

**Comment by ceuk** [comment]
here's a little bash script using fzf, bat and ripgrep which can be run as a task and works nicely:

```bash
#!/usr/bin/env bash

fzf --layout=default --preview-window=up:60%:wrap \
  --disabled \
  --ansi \
  --bind "change:reload:sleep 0.1; rg --column --line-number --no-heading --color=always --smart-case {q} || true" \
  --bind "ctrl-d:preview-down,ctrl-u:preview-up" \
  --bind "enter:execute-silent(zed {1}:{2})+abort" \
  --delimiter ':' \
  --with-nth='${1}' \
  --preview '
    FILE={1}
    LINE={2}
    if [[ -n "$FILE" && -n "$LINE" && -f "$FILE" ]]; then
      bat --style=numbers,plain --color=always --theme=tokyonight --highlight-line {2} {1}
    else
      echo ""
    fi
  ' \
  --preview-window '+{2}+4/3,<80(up),border-bottom' \
  --scrollbar ' ' \
  --phony \
  --query "" \
  --prompt "Search > " \
  --multi
```

Example:

![Image](https://github.com/user-attachments/assets/8f56151f-b542-4f8b-8cdd-4e5d61dd976d)

**Comment by nazriel** [comment]
Using https://github.com/alexpasmantier/television is quite nice workaround for having telescope like UI for finding files and greping (space space and space / bindings <3). 

The only missing bit I am having is the ability to pick open buffers in the similar fashion. 
I wonder if somebody came up with a solution for that?
Is there a way to programatically list open buffers in zed somehow?

**Comment by baggiponte** [comment]
Appreciate the suggestions - might be overzealous but I would like to add my +1 for this feature. I've been using neovim for 3+ years and while the experience so far with Zed is positive, this is definitely a big missing feature. Live grep search is quite important and will combo really well with the multi-file edit feature 😊

**Comment by yara-blue** [comment]
Telescope serves three goals for me:
- Navigation: fuzzy search is faster & easier then clicking in a file tree
- Collecting: example, getting a list of items to apply a (vim) macro to
- Exploration: example, find a function by a word in its doc comment

This issue has identified fast iteration (real time search) and fuzziness as key however I would like to zoom in on context. That is what makes telescope unique for *exploration*.

Telescope provides context in both the result list and the preview. When searching for a string in all project files (the most common telescope use-case) the context is:
- Showing the path to it next to each result.
- Showing the lines around the selected search result. Great for local context like function arguments, return type etc.

For other search dimensions (like searching through all LSP symbols) these end up slightly different. Note there are a lot of search dimensions like git commits, tabs or even custom searches added by the user in lua.

Now ill get to my point :)

Lets build upon telescope and add a component to zed which offers the ideal context in a real time fuzzy search tool. The context type can change depending on what kind of result we are presenting. 

As an example: searching for a string in the project. 
- A result inside code should show a few surrounding lines and relevant structure further away. Things like impl blocks/loops. That should be doable using tree-sitter (See [treesittes-context](https://github.com/nvim-treesitter/nvim-treesitter-context)). 
- Long paths can be trimmed with some heuristic to show the interesting differences. 
- Display the date a line was last edited and by whom under the preview? 

There are a lot of options to make code exploration even better, if a telescope like component becomes part of Zed lets architect it such that we can keep adding more context.

**Comment by gustavotrott** [comment]
Any update in this regard?
I think about this issue every day whenever I need to search for something in the project 😅

**Comment by ceuk** [comment]
I really recommend [this script](https://github.com/zed-industries/zed/issues/8279#issuecomment-2843477959) it's still what I use and the closest thing to Telescope I've found or come up with so far except television which is also great

https://github.com/user-attachments/assets/3dc3406c-8e52-40ba-a522-ddb5db3f34a1

**Comment by justsomelegs** [comment]
i've wanted this kind of feature for a while and thought i would take a crack at a first iteration

its early doors but heres a _non-prefect_ working 1st attempt - still alot to get right and optimise but its working atleast i suppose

https://github.com/user-attachments/assets/2de1eade-d2e7-44ef-bdb9-6087e7bb77cf

**Comment by Dima-369** [comment]
> its early doors but heres a non-prefect working 1st attempt - still alot to get right and optimise but its working atleast i suppose

I wonder if you can push the code to your fork; it has no commits pushed yet? I am always interested in seeing the code for such features. Thanks 🙂

**Comment by phibr0** [comment]
I really like how Fleet handled this, their default "goto" switcher handles both filenames and symbols while showing the file contents 

https://www.jetbrains.com/help/fleet/1.33/navigation.html#goto

**Comment by portseif** [comment]
Php Storm also has this and is called `Search Everywhere`: https://www.jetbrains.com/help/phpstorm/2025.3/searching-everywhere.html

[EDIT]
Created this today: https://github.com/zed-industries/zed/pull/44674

![Image](https://github.com/user-attachments/assets/ed5ec2c3-6342-4f98-b184-368df0e4f3eb)

**Comment by justsomelegs** [comment]
been busy for the last week at my day job but started working on my _attempt_ some more:

started redesigning it in a modular way so that more contexts can be added and such:

https://github.com/user-attachments/assets/e8d4cf93-217c-4431-81a2-d08e823abf8a


still a lot to get right but i think its going in a good direction :) gonna see if i get some more bugs ironed out then make a PR tonight hopefully

**Comment by ceuk** [comment]
@justsomelegs genuinely the coolest file search GUI I've seen in any IDE I think. If I ever bump into you IRL I will buy you a pint (or some trousers...). Really great work

**Comment by justsomelegs** [comment]
> [@justsomelegs](https://github.com/justsomelegs) genuinely the coolest file search GUI I've seen in any IDE I think. If I ever bump into you IRL I will buy you a pint (or some trousers...). Really great work

thank you :)



just got git status preview working on search results, as well as streaming results in.

not gonna be feature rich the intial PR but i think it has a lot of potential to be something cool that i'm enjoying working on :)

**Comment by yara-blue** [comment]
> not gonna be feature rich the intial PR

If you can please get a draft up so we can discuss design & implementation with you :). Make sure to tag me in it!

I'd love to support this

**Comment by justsomelegs** [comment]
> If you can please get a draft up so we can discuss design & implementation with you :). Make sure to tag me in it!


thank you, will do just got back from dinner, will get it up soon for you :)

**Comment by baldwindavid** [comment]
@justsomelegs - Looks cool. Just so you're aware, there is also #44530 in flight.

**Comment by justsomelegs** [comment]
just got my draft PR up, a lot still left to clean up as said before but its there now if people want to have a look around #44942 

@yara-blue

**Comment by viktaur** [comment]
Any updates on this?

**Comment by ryanwinchester** [comment]
> Any updates on this?

#46478

**Comment by aizigao** [comment]
for zed 1.9.0  I can use cmd+p than cmd+alt+p to trigger new scopeStyle picker, but how to binding it to cmp+p directly 😃 

<img width="2330" height="1300" alt="Image" src="https://github.com/user-attachments/assets/f054adfb-b2e5-4fe1-a345-6a015ff5a7d7" />

**Comment by lixiaoyan** [comment]
It would be ideal to have configuration options to control whether the preview is shown by default and the direction of the split.

**Comment by lukasbash** [comment]
> It would be ideal to have configuration options to control whether the preview is shown by default and the direction of the split.

I would like to work on this!

**Comment by iggyzuk** [comment]
> > It would be ideal to have configuration options to control whether the preview is shown by default and the direction of the split.
> 
> I would like to work on this!

It appears to have been implemented in the latest preview (1.10.0)

https://zed.dev/releases/preview

https://github.com/zed-industries/zed/pull/59849

**Comment by youvegotmoxie** [comment]
> It appears to have been implemented in the latest preview (1.10.0)
> 
> https://zed.dev/releases/preview
> 
> [#59849](https://github.com/zed-industries/zed/pull/59849)

If I'm reading that PR right it looks like you'd need to configure how the picker preview is displayed for each project. That behavior seems less than ideal.

**Comment by lukasbash** [comment]
Sorry if I am mistaken (it is my first PR on this project) but I think #59849 is a different thing. Doesn't it make the text finder remember your last query/filters so reopening lands on the same results? So the ask (preview shown by default and where it's placed) is still open.

**Comment by 0x6e697678** [comment]
Do I miss anything ? I can't find a way to open the search modal. Still have to use the search pane for now. It seems that the `project_search::OpenTextFinder` command is missing.

**Comment by kaineer** [comment]
> Do I miss anything ? I can't find a way to open the search modal. Still have to use the search pane for now. It seems that the `project_search::OpenTextFinder` command is missing.

`text_finder::Toggle` ftw

**Comment by Koshkaj** [comment]
I thought this functionality would have worked with references as well, but references still opens up a new tab where you see all the places of a specific thing that you are searching for. 

Also, why do you need to specifically press preview afterwards once you've searched. and at the preview window, how can I set to display only file names in the left pane? (like Telescope)

I am wondering how the PR that got merged closed this ticket, since it seems that functionality is far off from the original request.

**Comment by dbkegley** [comment]
> I thought this functionality would have worked with references as well, but references still opens up a new tab where you see all the places of a specific thing that you are searching for.

There's a separate issue tracking pickers for LSP results https://github.com/zed-industries/zed/issues/59829

**Comment by 0x6e697678** [comment]
> > Do I miss anything ? I can't find a way to open the search modal. Still have to use the search pane for now. It seems that the `project_search::OpenTextFinder` command is missing.
> 
> `text_finder::Toggle` ftw

Many thanks. I skimmed the PR and only found the `project_search::OpenTextFinder`.

**Comment by viktaur** [comment]
What's the state of this feature? Is it going to be available on any preview or stable release?
